### PR TITLE
reader_concurrency_semaphore: don't evict inactive readers needlessly

### PR DESCRIPTION
**Pull request by denesb**
Inactive readers should only be evicted to free up resources for waiting readers. Evicting them when waiters are not admitted for any other reason than resources is wasteful and leads to extra load later on when these evicted readers have to be recreated end requeued. This patch changes the logic on both the registering path and the admission path to not evict inactive readers unless there are readers actually waiting on resources.
A unit-test is also added, reproducing the overly-agressive eviction and checking that it doesn't happen anymore.

Fixes: #11803

### CODE REVIEW

**Comment by denesb** [comment]
Don't merge yet, I want to extend the tests a bit, there will be a v2.

**Comment by denesb** [comment]
New in v2:
* Extend tests:
    - extend the existing negative tests
    - add positive tests too
* Don't evict inactive reads when resources are right on the limit, but there is no demand (found by extended tests).

**Comment by denesb** [comment]
New in v2.1:
* Added a thread::yield() to allow debug build to schedule fiber evicting inactive reads again.

**Review by avikivity — APPROVED** [review]
Tricky stuff

**Comment by denesb** [comment]
Ping.

**Comment by denesb** [comment]
Ping.

File: test/boost/reader_concurrency_semaphore_test.cc
Comment (tchaikov) [inline test/boost/reader_concurrency_semaphore_test.cc:1663]: in #13255, we [changed](https://github.com/scylladb/scylladb/pull/13255/commits/156e5d346d427abc4c79075cb2e71d68ed35d703#diff-46ce50f665ed19eaed28937148d633880d420c375fa7fbed2cb9261047a57ca9R389) the function signature to pass `trace_state` down to semaphore, so might want to update this caller accordingly.

**Review by tchaikov — CHANGES_REQUESTED** [review]

**Comment by denesb** [comment]
Dequeued, breaks the build.

**Comment by denesb** [comment]
Rebased and fixed the build.

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/381/

https://github.com/scylladb/scylladb/issues/13211

Re-kicked

**Comment by denesb** [comment]
Ping.

**Comment by yaronkaikov** [comment]
@denesb @avikivity Can we merge this?

**Comment by denesb** [comment]
> @denesb @avikivity Can we merge this?

I cannot merge my own PR. We need another @scylladb/scylla-maint maintainer to merge it.

**Comment by denesb** [comment]
@scylladb/scylla-maint ping. This PR is holding up a release, for weeks now.

**Comment by nyh** [comment]
@avikivity already approved it, so I'll merge. I hope the CI run is not stale :-(

**Comment by denesb** [comment]
> @avikivity already approved it, so I'll merge. I hope the CI run is not stale :-( (I'll verify it still compiles before merging).

Thanks!

**Pull request closed without merging**
